File: app/src/main/java/com/fernandosierra/door2door/presentation/screens/splash/SplashPresenter.java
package com.fernandosierra.door2door.presentation.screens.splash;

import android.support.annotation.NonNull;

import com.fernandosierra.door2door.domain.interactor.SetupInteractor;
import com.fernandosierra.door2door.presentation.base.BasePresenter;
import com.fernandosierra.door2door.presentation.util.Door2DoorUtils;
import com.fernandosierra.door2door.presentation.util.observer.CompletableObserverAdapter;

import javax.inject.Inject;

public class SplashPresenter extends BasePresenter<SplashView> {
    private static final String LOCAL_JSON = "local.json";
    private final SetupInteractor setupInteractor;

    @Inject
    public SplashPresenter(@NonNull SplashView splashView, @NonNull SetupInteractor setupInteractor) {
        super(splashView);
        this.setupInteractor = setupInteractor;
    }

    public void init() {
        secureViewCall(SplashView::init);
    }

    public void loadLocalData() {
        secureViewCall(SplashView::showLoader);
        final String localDataJson = Door2DoorUtils.instance.readJsonAsset(LOCAL_JSON);
        if (localDataJson != null) {
            setupInteractor.setupFromLocalData(localDataJson, new CompletableObserverAdapter() {
                @Override
                public void onComplete() {
                    secureViewCall(view -> view.hideLoader(() -> secureViewCall(SplashView::goRoutes)));
                }

                @Override
                public void onError(Throwable e) {
                    secureViewCall(view -> view.hideLoader(() -> secureViewCall(SplashView::showErrorDialog)));
                }
            });
        } else {
            secureViewCall(view -> view.hideLoader(null));
        }
    }
}

File: app/src/main/java/com/fernandosierra/door2door/presentation/screens/splash/SplashPresenter.kt
package com.fernandosierra.door2door.presentation.screens.splash

import com.fernandosierra.door2door.domain.interactor.SetupInteractor
import com.fernandosierra.door2door.presentation.base.BasePresenter
import com.fernandosierra.door2door.presentation.util.readJsonAsset
import kotlinx.coroutines.experimental.android.UI
import kotlinx.coroutines.experimental.async
import kotlinx.coroutines.experimental.launch
import javax.inject.Inject

class SplashPresenter @Inject constructor(view: SplashView, private val setupInteractor: SetupInteractor)
    : BasePresenter<SplashView>(view) {
    companion object {
        private const val LOCAL_JSON = "local.json"
    }

    fun init() = view { it.init() }

    fun loadLocalData() {
        view { it.showLoader() }

        launch(UI) {
            val deferred = async {
                var localDataJson: String? = null
                view { localDataJson = it.ctx?.readJsonAsset(LOCAL_JSON) }
                if (localDataJson != null) {
                    setupInteractor.setupFromLocalData(localDataJson ?: "")
                } else {
                    throw RuntimeException("Couldn't load data from local JSON")
                }
            }
            deferred.join()
            if (deferred.getCompletionExceptionOrNull() == null) {
                view { it.goRoutes() }
            } else {
                view { it.showErrorDialog() }
            }
        }
    }
}